hal/spi_api.h

```diff
@@ -19,220 +18,211 @@
 #ifndef MBED_SPI_API_H
 #define MBED_SPI_API_H
 
+#include <stdbool.h>
+#include <stdint.h>
 #include "device.h"
-#include "hal/dma_api.h"
-#include "hal/buffer.h"
 
 #if DEVICE_SPI
-
-#define SPI_EVENT_ERROR       (1 << 1)
-#define SPI_EVENT_COMPLETE    (1 << 2)
-#define SPI_EVENT_RX_OVERFLOW (1 << 3)
-#define SPI_EVENT_ALL         (SPI_EVENT_ERROR | SPI_EVENT_COMPLETE | SPI_EVENT_RX_OVERFLOW)
-
-#define SPI_EVENT_INTERNAL_TRANSFER_COMPLETE (1 << 30) // Internal flag to report that an event occurred
-
-#define SPI_FILL_WORD         (0xFFFF)
-#define SPI_FILL_CHAR         (0xFF)
-
-#if DEVICE_SPI_ASYNCH
-/** Asynch SPI HAL structure
- */
-typedef struct {
-    struct spi_s spi;        /**< Target specific SPI structure */
-    struct buffer_s tx_buff; /**< Tx buffer */
-    struct buffer_s rx_buff; /**< Rx buffer */
-} spi_t;
-
-#else
-/** Non-asynch SPI HAL structure
- */
-typedef struct spi_s spi_t;
-
-#endif
-
-#ifdef __cplusplus
-extern "C" {
-#endif
-
 /**
- * \defgroup hal_GeneralSPI SPI Configuration Functions
+ * \defgroup hal_spi SPI: Serial peripheral interface HAL API.
+ * Low level interface to the serial peripheral interface of a target.
+ *
+ * A SPI peripheral might be used by multiple spi_t as it only represents a communication
+ * channel towards a slave (or a master).
+ * Two SPI instances using the same peripheral can be identified by their SS pin.
+ *
+ * # Defined behaviour
```

Comment:
Would it be possible to define some default configuration for each SPI target? There are many config options like: `clock_frequency`, `master/slave`, `clock_phase`, `polarity`, `word_length`, `mode`. It might be problematic from testing perspective.
I assume that if device supports SPI, then initialization status should be `SPI_RESULT_OK` for at least one of all the possible configurations. I could check all of them, but there could be problem with `clock_frequency` which doesn't have predefined values.
I was wondering if we could request vendors to define default SPI config supported for their platform for master and slave (I assume that some platforms may provide only slave mode?).

I suggest to add something like:

```
WEAK spi_init_s * get_default_spi_config(bool is_master)
{
    if (is_master) {
        return NULL;  // Vendor shell provide pointer to the default master SPI config
    } else {
        return NULL; // Vendor shell provide pointer to the default slave SPI config
    }
}

```
This function should be overridden by the target specific driver. This also would be help for the user to make a quick start.

What do you think?

Reply:
We could test all configurations if we could get the capabilities of the target from the vendor.
This is something we discussed with @c1728p9 but we are not settled yet. 

We could require vendors to provide a `SPI_DEFAULT_FRENQUENCY` but it would still leave in the worst case more than 5000 possible configurations to be tested.

Which of the following are acceptable :
- to test against 1 configuration and assume all others are supported ?
- to require that all configurations must be supported except for `spi_modes_t`, `word_length` and `continuous mode` which would be obtained through a capability request ?
  The tests would pick randomly a configuration in the range of the capabilities and required options ?

Should we also constraint min/max transaction size ?

Reply:
I think that capability list is the best option. But in such case we should also define default frequency and word length.
Regarding functional tests in loop-back mode I think that we should not check all possibilities, but limit tests to the following:
- only default frequency,
- most significant bit first / most significant bit last (if target support both settings),
- two different fill_symbols,
- only default word_length,
- all supported modes - I don't know yet if all can be tested in loop-back mode (if target support all settings).

I'm not familiar with `clock_phase` and `clock_polarity`, but if target supports both options (`true`, `false`), then also this should be tested.

What do you mean by "min/max transaction size" ? `tx_count/rx_count` in `spi_transfer_args_t`?

Reply:
Yes, `tx_count` and `rx_count` are in `[0, UINT32_MAX]`.

It is very unlikely that any device support `UINT32_MAX` due to ram limitation but we should take care that hal impl do actually support transactions of 0, 1, n symbol(s).

Reply:
My concerns with these solutions are :
- Using a single configuration : we are obviously not covering many feature of our HAL.
- Using *randomly picked* configuration : we potentially introduce test inconsistency where without changing code we might get different test results as a defect would only be detected "by chance".

Reply:
I agree with your concerns. That is why we should select reasonable number of test cases which cover most of HAL features. Testing everything is impossible, additionally this functional loop-back tests will be deactivated on CI - mainly for driver developers to check if their drivers meets the requirements. I think it is possible to test everything what is needed if we have device SPI capability list.

Regarding "min/max transaction size" from testing perspective I can't see a reason to test large transfers (i.e. up to memory limit). Test should check only few different sizes like 0 (robust case), 1 and N symbols, where N is some reasonable value greater than 1.